### PR TITLE
17 app channel tests 

### PR DESCRIPTION
**Pull request by Joe-Dunleavy**
Pull request for issue #17 App Channel tests.

image

### CODE REVIEW

**Comment by AMehtaSL** [comment]
> * The first test on the list has to run first due App A having to broadcast context to the same channel app B has joined after each test so that app B can close itself. Running the test further down in the list will cause the test to receive this context when running `testChannel.getCurrentContext()`, rendering the test invalid.
> 

I don't think we can guarantee the order in which tests run. In any case even if we can, it is a bad practice to have individual tests rely on the order of running. I don't understand the problem, maybe we can have a call about it?

**Comment by Joe-Dunleavy** [comment]
@AMehtaSL I agree it's not ideal. I'm not sure it's possible to resolve the issue without removing the functionality for shutting down the mock apps though. And keeping those mock apps open introduces further issues. happy to give you a call to show you the problem

File: mock/channels/index.html [hunk 1]
```diff
@@ -1,73 +1,138 @@
...
-        const broadcast = async (contextBroadcasts) => {
-          for (const systemBroadcast in contextBroadcasts) {
-            if (contextBroadcasts[systemBroadcast]) {
-              await window.fdc3.broadcast({
-                "type": `fdc3.${systemBroadcast}`
-              });
+      //Channels app joins test-channel
+      const joinAppChannel = async () => {
+        try {
+          var appChannel = await fdc3.getOrCreateChannel("test-channel");
```
Comment (AMehtaSL) [inline mock/channels/index.html:84]: Can use const instead of var

File: mock/channels/index.html [hunk 2]
```diff
@@ -1,73 +1,138 @@
...
+
+      //Channels app joins channel 1
+      const joinSystemChannel = async () => {
+        const channels = await window.fdc3.getSystemChannels();
+        if (channels.length > 0) {
+          try {
+            await window.fdc3.joinChannel(channels[0].id);
+            document.getElementById("context").innerHTML += `Joined user channel ${channels[0].id}/ `;
+          } catch (ex) {
+            throw new Error("Error while calling joinChannel in app B: " + (ex.message ?? ex))
```
Comment (AMehtaSL) [inline mock/channels/index.html:74]: Does throwing errors like this achieve anything i.e. does it get displayed in the app? I would assume it doesn't do anything and might be better to add to the document?

File: tests/src/test/fdc3.broadcast.ts [hunk 3]
```diff
@@ -13,272 +13,554 @@ export default () =>
...
     };
 
-    afterEach(async () => {
-      if (listener !== undefined) {
-        await listener.unsubscribe();
-        listener = undefined;
+    beforeEach(async () => {
+      await unsubscribeListeners();
+      if (channelsAppContext.joinAppChannel === false) {
+        await window.fdc3.leaveCurrentChannel();
```
Comment (AMehtaSL) [inline tests/src/test/fdc3.broadcast.ts:23]: I don't see the harm in just running this unconditionally?

File: tests/src/test/fdc3.broadcast.ts [hunk 4]
```diff
@@ -13,272 +13,554 @@ export default () =>
...
+
+          assert.isObject(listener);
+          expect(typeof listener.unsubscribe).to.be.equals("function");
+
+          //App B joins the same app channel as A then broadcasts context
+          await window.fdc3.open("ChannelsApp", channelsAppContext);
+
+          function checkIfBothContextsReceived() {
+            contextsReceived++;
+            if (contextsReceived > 1) {
```
Comment (AMehtaSL) [inline tests/src/test/fdc3.broadcast.ts:429]: I'm not sure about this. I'd prefer an expectation for a precise number of contexts or even better storing a list of everything received and then verifying it is as expected

File: tests/src/test/fdc3.broadcast.ts (hunk 3 above)
Comment (Joe-Dunleavy) [reply tests/src/test/fdc3.broadcast.ts:23]: it doesn't do anything when running app channel tests. It only works on user channels. But I guess I could just leave it anyway

File: mock/channels/index.html (hunk 2 above)
Comment (Joe-Dunleavy) [reply mock/channels/index.html]: It doesn't. I was going to do this but forgot

File: mock/channels/index.html [hunk 5]
```diff
@@ -1,73 +1,138 @@
...
+            "closeWindow",
+            () => {
+              closeFinsembleWindow();
+            });
+        }
+      }
+
+      const joinChannel = async (useAppChannel) => {
+        if (useAppChannel) {
+          return await joinAppChannel();
```
[the remark this answers is not in the document]
Comment (Joe-Dunleavy) [reply mock/channels/index.html:48]: it's returning the channel object:

image

File: mock/channels/index.html (hunk 1 above)
Comment (Joe-Dunleavy) [reply mock/channels/index.html]: Not sure why I did that. I must have momentarily lapsed into c# without realising

**Review by AMehtaSL — APPROVED** [review]

**Pull request merged**